caffe2/sgd/adadelta_op_gpu.cu

```diff
@@ -0,0 +1,168 @@
+    int N,
+    const float* w,
+    const float* g,
+    const float* h,
+    const float* d,
+    float* nw,
+    float* nh,
+    float* nd,
+    float epsilon,
+    float decay,
+    const float* lr,
+    CUDAContext* context) {
+  AdadeltaUpdate<<<
+      CAFFE_GET_BLOCKS(N),
+      CAFFE_CUDA_NUM_THREADS,
+      0,
+      context->cuda_stream()>>>(N, w, g, h, d, nw, nh, nd, epsilon, decay, lr);
+}
+
+template <typename SIndex, typename THalf>
+__global__ void SparseAdadeltaKernel(
+    const size_t N,
+    const size_t grad_slice_sz,
+    const float epsilon,
+    const float decay,
+    THalf* param,
+    THalf* param_mom,
+    THalf* param_mom_delta,
+    const SIndex* indices,
+    const float* grad,
+    const float* lr) {
+  const float LR = lr[0];
+  CUDA_1D_KERNEL_LOOP(i, N) {
+    const size_t gradIdx = i;
+    const SIndex index = indices[i / grad_slice_sz];
+    const size_t paramIdx = index * grad_slice_sz + (i % grad_slice_sz);
+
+    float mom_new = mixed_mult(decay, param_mom[paramIdx]) +
+        (1.0f - decay) * grad[gradIdx] * grad[gradIdx];
+    mixed_store(&mom_new, &(param_mom[paramIdx]));
+    float grad_new = (sqrt(mixed_add(epsilon, param_mom_delta[paramIdx])) /
+                      sqrt(mom_new + epsilon)) *
+        grad[gradIdx];
+    float param_new = mixed_add(LR * grad_new, param[paramIdx]);
+    mixed_store(&param_new, &(param[paramIdx]));
+    float mom_delta_new = mixed_mult(decay, param_mom_delta[paramIdx]) +
+        (1.0f - decay) * grad_new * grad_new;
+    mixed_store(&mom_delta_new, &(param_mom_delta[paramIdx]));
+  }
+}
+
+template <typename T, class Context>
+class CUDASparseAdadeltaOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+  CUDASparseAdadeltaOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws),
+        epsilon_(OperatorBase::GetSingleArgument<T>("epsilon", 1e-5f)),
+        decay_(OperatorBase::GetSingleArgument<T>("decay", 0.95f)) {}
+
+  bool RunOnDevice() override {
+    // Enforce shapes
+    CAFFE_ENFORCE_EQ(Input(PARAM).size(), Input(MOMENT_GRAD).size());
+    CAFFE_ENFORCE_EQ(Input(PARAM).size(), Input(MOMENT_DELTA).size());
+    CAFFE_ENFORCE_EQ(Input(LR).size(), 1);
+    CAFFE_ENFORCE_EQ(
+        Input(PARAM).size_from_dim(1),
+        Input(GRAD).size_from_dim(Input(INDICES).ndim()));
+    // Enforce domain constraints on attributes
+    CAFFE_ENFORCE_GE(static_cast<T>(epsilon_), 0.0f);
+    CAFFE_ENFORCE_GT(static_cast<T>(decay_), 0.0f);
+    CAFFE_ENFORCE_LT(static_cast<T>(decay_), 1.0f);
+
+    return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(
+        this, Input(INDICES));
+  }
+
+  template <typename IndexType>
+  bool DoRunWithType() {
+    auto n = Input(INDICES).size();
+    if (n == 0) {
+      return true;
+    }
+    return DispatchHelper<TensorTypes2<float, float16>, IndexType>::call(
+        this, Input(PARAM));
+  }
+
+  template <typename IndexType, typename THalf>
```

Comment:
I'm a little confused about the template here. Since on the Op-level we have a template type T, how is the relationship between T and THalf. I think it's better to remove T at Op-level if you want to use DoRunWithType.

Reply:
For your comments regarding T and THalf, since we don't have kernels for other types like float16 or double, I have removed the typename T at Op-level, for simplicity. However, even when we had type T at the Op level, I don't think there would be an issue.

Let me make sure that what I understand is reasonable (so that we are on the same page). I think this is what is happening with the current implmentation:
- Since DispatchHelper in L109 (`return DispatchHelper<TensorTypes<int32_t, int64_t>>::call(this, Input(INDICES))`, would produce something the nested if-else code like the following:
```
if (this->INPUT(INDICES).template IsType<int32_t>()) {
        DoRunWithType<int32_t>();
    } else if (this->INPUT(INDICES).template IsType<int32_t>()) {
        DoRunWithType<int64_t>();
    }
```
Now we have `DispatchHelper<TensorTypes2<float, float16>` in the `DoRunWithType()`. This again will call `DoRunWithType2()` with `<float>` or `<float16>` depending on the type of `Input(PARAM)`.

So even when we had type T at the Op level, only `const auto* gradIn = Input(GRAD).template data<T>();` is different (L127). So this would have been still fine.

Reply:
Currently it looks much better, thanks.